File: LeetCode/98 - Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    bool isValidBST(TreeNode *root)
    {
        if (!root)
            return true;

        stack<int> s;
        dfs(root, s);
        int pre = s.top();
        s.pop();
        while (!s.empty()) {
            int cur = s.top();
            s.pop();
            if (cur >= pre)
                return false;
            pre = cur;
        }
        return true;
    }

    void dfs(TreeNode *root, stack<int> &s)
    {
        if (!root)
            return;
        dfs(root->left, s);
        s.push(root->val);
        dfs(root->right, s);
    }
};

File: LeetCode/grind75/98 - Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

import math
from typing import Optional


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.dfs(root, -math.inf, math.inf)

    def dfs(self, root, lower_bound, upper_bound) -> bool:
        if not root:
            return True

        if root.val <= lower_bound or root.val >= upper_bound:
            return False

        return self.dfs(root.left, lower_bound, root.val) and self.dfs(
            root.right, root.val, upper_bound
        )
